snapcraft/plugins/v2/colcon.py

```diff
@@ -0,0 +1,207 @@
+                    "minitems": 1,
+                    "items": {"type": "string"},
+                    "default": [],
+                },
+                "colcon-catkin-cmake-args": {
+                    "type": "array",
+                    "minitems": 1,
+                    "items": {"type": "string"},
+                    "default": [],
+                },
+                "colcon-ament-cmake-args": {
+                    "type": "array",
+                    "minitems": 1,
+                    "items": {"type": "string"},
+                    "default": [],
+                },
+                "colcon-packages-ignore": {
+                    "type": "array",
+                    "minitems": 1,
+                    "uniqueItems": True,
+                    "items": {"type": "string"},
+                    "default": [],
+                },
+            },
+        }
+
+    def get_build_snaps(self) -> Set[str]:
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        return {
+            "python3-colcon-common-extensions",
+            "python3-rosdep",
+            "python3-wstool",
+            "python3-rosinstall",
+        }
+
+    def get_build_environment(self) -> Dict[str, str]:
+        return {
+            "AMENT_PYTHON_EXECUTABLE": "/usr/bin/python3",
+            "COLCON_PYTHON_EXECUTABLE": "/usr/bin/python3",
+            "ROS_PYTHON_VERSION": "3",
+        }
+
+    def _get_colcon_build_command(self) -> str:
+        cmd: List[str] = [
+            "colcon",
+            "build",
+            "--merge-install",
+            "--install-base",
+            "$SNAPCRAFT_PART_INSTALL",
+        ]
+
+        if self.options.colcon_packages_ignore:
+            cmd.extend(["--packages-ignore", *self.options.colcon_packages_ignore])
+
+        if self.options.colcon_packages:
+            cmd.extend(["--packages-select", *self.options.colcon_packages])
+
+        if self.options.colcon_ament_cmake_args:
+            cmd.extend(["--ament-cmake-args", *self.options.colcon_ament_cmake_args])
+
+        if self.options.colcon_catkin_cmake_args:
+            cmd.extend(["--catkin-cmake-args", *self.options.colcon_catkin_cmake_args])
+
+        # Specify the number of workers
+        cmd.extend(["--parallel-workers", "${SNAPCRAFT_PARALLEL_BUILD_COUNT}"])
+
+        return " ".join(cmd)
+
+    def get_build_commands(self) -> List[str]:
+        return [
+            ". /opt/ros/$ROS_DISTRO/setup.sh",
+            "if [ ! -f /etc/ros/rosdep/sources.list.d/20-default.list ]; then sudo rosdep init; fi",
+            "rosdep update --rosdistro $ROS_DISTRO",
+            "rosdep install --from-paths . -y -r",
+            self._get_colcon_build_command(),
+            f"{sys.executable} -m snapcraft.plugins.v2.colcon stage-runtime-dependencies",
+        ]
+
+
+@click.group()
+def plugin_cli():
+    pass
+
+
+@plugin_cli.command()
+@click.option("--build-base", envvar="SNAPCRAFT_BUILD_BASE", required=True)
```

Comment:
Considering this plugin is strictly for core20, one could _hardcode_ the base here.

Reply:
funny i didn't consider that, lol.  it could also be exported by the extension I suppose, but eventually other bases will come along anyways.

Reply:
Is it? Dashing is for core18, will that cause issues?

Reply:
this is actually foxy, I just did my prototyping with dashing because foxy it was not yet ready...